### ISSUE TITLE
NIDM view rounds coordinates

### ISSUE THREAD

**Issue by chrisfilo** [NeuroVault/NeuroVault#371]
This might be a papaya issue.

**Comment by vsoch** [comment]
I think this must be papaya, because the coordinates from the query are float, and I [parse as float](https://github.com/vsoch/nidmviewer/blob/master/js/script.js#L5). It would be ideal to not have rounding, but it seems like a small issue.

**Comment by chrisfilo** [comment]
Maybe @rii-mango can chime in.

The issue is that coordinates are rounded to a multiplier of 2 (since the background image is 2x2x2mm). This sometimes results in confusing behavior when user clicks on a coordinate and the viewer shows an empty slice. (see here http://neurovault.org/collections/877/fsl_course_av.nidm# especially the second contrast in this NIDM file). This is what @satra was referring to in one of his emails.

**Comment by vsoch** [comment]
holy crap!! I didn't realize that it was deployed! Wicked!! :D

**Comment by rii-mango** [comment]
I think I understand the issue.  At least, I see how Papaya may be truncating one end of the image when it could be interpolating some of those bottom slices.  Try this:

In `transform.js`, in function `papaya.volume.Transform.prototype.getVoxelAtCoordinate()`, comment this out:
```
    //if ((xTrans < 0) || (xTrans >= this.imageDimensions.xDim) || (yTrans < 0) ||
    //    (yTrans >= this.imageDimensions.yDim) || (zTrans < 0) || (zTrans >= this.imageDimensions.zDim)) {
    //    return 0;
    //}
```

In `voxelvalue.js`, in the "constructor", add this line:
```
    this.interpFirstPass = [[0, 0], [0, 0]];
    this.interpSecondPass = [0, 0];
    this.offsetMax = this.imageData.data.length - 1;  // add this line
};
```

In the same file, in `papaya.volume.VoxelValue.prototype.getVoxelAtOffset()`, add this:
```
papaya.volume.VoxelValue.prototype.getVoxelAtOffset = function (volOffset, timepoint) {
    var dataScaleIndex,
        offset = volOffset + (this.volSize * timepoint);

    if ((offset > this.offsetMax) || (offset < 0)) {  // add this block
        return 0;
    }

    if (this.usesGlobalDataScale) {
        return (this.checkSwap(this.imageData.data[offset]) * this.globalDataScaleSlope) +
            this.globalDataScaleIntercept;
    } else {
        dataScaleIndex = parseInt(offset / this.sliceSize);
        return (this.checkSwap(this.imageData.data[offset]) * this.dataScaleSlopes[dataScaleIndex]) +
            this.dataScaleIntercepts[dataScaleIndex];
    }
};
```

This should at least let it interpolate those bottom slices, averaging in zeros when it's outside the image bounds.

**Comment by vsoch** [comment]
Awesome! I will give this a try and report back. We will definitely want to be precise about the locations in NeuroVault. The coordinates are essential because the next step is to be able to share a neurovault image and send a particular coordinate to look at. For example, this url:

      http://neurovault.org/collections/877/fsl_course_av.nidm

Might become this URL:

      http://neurovault.org/collections/877/fsl_course_av.nidm?x=10.5&y=11&z=12.2

And that would be incredibly powerful for sharing and communicating about maps, but also learning how people are viewing them. We can accomplish this with the current functionality (likely with some kind of share button) but in the same way papaya can expect url parameters, it might be something to add so that it is intrinsic to papaya and other users could take advantage of the functionality.

I should have some time this weekend to test your suggestion, thank you so much! :O)

**Comment by vsoch** [comment]
Ok, I made the modifications and then tested the viewer with the new and old version, shown below:

![image](https://cloud.githubusercontent.com/assets/814322/11023834/18830212-8638-11e5-89f5-dc6ecf612bf5.png)

The picture on the left is the new (updated) version with the code suggested by @rii-mango , and the picture on the right is the current (old) version.  While the coordinate shown in the viewer is still truncated, the display appears to be showing a coordinate that is better fit to the center of the activation cluster. Note that while you can't see it in the first image, the display equally shows -4, 80, -14 for the coordinate  -3.77 -80.4 -13.0. I'm not sure why the Z coordinate would jump to -14 (it might be wrong entirely), although the display does look improved. I have made available both so that @chrisfilo can take a look!

- [old version](http://vbmis.com/bmi/share/papaya/coords/old.html)
- [new version](http://vbmis.com/bmi/share/papaya/coords/)

@chrisfilo If this is a suitable fix, I can update the papaya in the nidmviewer to make this change. @rii-mango, if you want to integrate into the application to fix #36, I have the changed code as a [branch here](https://github.com/vsoch/Papaya/tree/fix/round_coords). We may want to delete the commented out part instead of commenting.

**Comment by nicholst** [comment]
Minor point, but if you want to be sure about localization (i.e. where the peak is) it's important to *not* use interpolation in the viewer (i.e. use nearest-neighbor), which in Papaya-ese is `smoothDisplay=false` (it defaults to true).

**Comment by chrisfilo** [comment]
I think the Z coordinate jumped to -14 because we use 2x2x2mm template as a
background (so everything needs to be a factor if two). I wonder how much
slower the rendering would be if we switched to a higher resolution
template (which would be bigger, but presumably browsers would cache it).

**Comment by nicholst** [comment]
PS: In our early playing with Papaya we were screwed up (I forget the details) until we set `worldSpace=true`... not sure if that is a factor here.

**Comment by vsoch** [comment]
@nicholst we set worldspace=true so we should be ok!

I just tried a version with a 1mm brain, and it's so slow that the timeout to let the brain map load is far too short (and thus the error). I can't tell if it's my shoddy internet connection of the T1, but I could go take a stroll around the block and build a birdhouse in the time it takes this map to load:

[go build a birdhouse](http://vbmis.com/bmi/share/papaya/mnimm/)

and here is one with a really long timeout, so at least the map loads (after 10 seconds!)

[maybe just time for a sandwich](http://vbmis.com/bmi/share/papaya/mnimm/old.html)

I also disabled the interpolation from this one.

**Comment by rii-mango** [comment]
I committed a fix in Papaya for the truncation when interpolating -- but note, this fix was very different than the one I proposed above.  The output is the same though.

One thing I wanted to point out is that the overlays (e.g., ZStatistic_T001) don't have origins set in their header.  Papaya is then defaulting to the center of the image array as origin.  It would probably be safer to specify a real origin in the overlay header.  (The MNI152 does have a real origin set.)  I noticed this when trying to compare the alignment between MNI152 and overlay in Papaya versus other viewers.

Let me know if you think there's any kind of alignment issue...

**Comment by satra** [comment]
@rii-mango - when you say origin, do you mean something other than what's in the affine transform in the nifti image? that transform should provide all the information necessary, if it is generated properly.

**Comment by rii-mango** [comment]
No, just the sform and qform transforms.  Here is what I see, for example (ZStatistic_T002), which puts the origin at the corner:

![screen shot 2015-11-09 at 10 44 47 am](https://cloud.githubusercontent.com/assets/1859709/11039862/45371cb6-86cf-11e5-876c-01469ef648e7.png)

Versus this one I made, which places the origin nearer to where I expect it:

![screen shot 2015-11-09 at 10 45 12 am](https://cloud.githubusercontent.com/assets/1859709/11039896/6e3192c2-86cf-11e5-9ae1-6703f9804918.png)

**Comment by vsoch** [comment]
Where and how was this nidm results produced?

**Comment by rii-mango** [comment]
It's this one: http://vbmis.com/bmi/share/papaya/coords/fsl/ZStatistic_T002.nii.gz

**Comment by vsoch** [comment]
Yes, that's one of the files from an NIDM-Results folder in NeuroVault NIDM showcase, this one I believe:

http://www.neurovault.org/collections/877/fsl_course_av.nidm.zip

I'm not sure how it's produced though.

**Comment by satra** [comment]
@rii-mango - that is indeed strange. those files have been stripped of origin info.
